File: packages/auditcore_kanban/src/auditcore_kanban/stats.py
"""Counts per column and progress (done column, else last column – audit_designer)."""

from __future__ import annotations

from .errors import JsonObject, JsonValue
from .model import Board, done_column


def percent(part: int, total: int) -> int:
    """Rounded percentage, half up (same as JavaScript ``Math.round``)."""
    if total <= 0:
        return 0
    return (200 * part + total) // (2 * total)


def board_stats(board: Board) -> JsonObject:
    """``total``, ``by_column``, ``done`` and ``progress`` (0–100)."""
    by_column: dict[str, JsonValue] = {}
    for column in board.columns:
        by_column[column.id] = sum(1 for c in board.cards if c.column_id == column.id)
    known = sum(1 for c in board.cards if board.column(c.column_id) is not None)
    done_id = done_column(board).id
    done = sum(1 for c in board.cards if c.column_id == done_id)
    return {
        "total": known,
        "by_column": by_column,
        "done_column": done_id,
        "done": done,
        "progress": percent(done, known),
    }


def checklist_progress(done: int, total: int) -> JsonObject:
    """Checklist counter of a card."""
    return {"done": done, "total": total, "percent": percent(done, total)}

File: packages-js/kanban-core/src/stats.ts
/** Zählung je Spalte und Fortschritt (Erledigt-Spalte, sonst letzte Spalte). */
import { doneColumn, findColumn, type Board } from './model'

/** Gerundeter Prozentwert, halbe Werte aufgerundet (= Math.round, wie stats.py). */
export function percent(part: number, total: number): number {
  if (total <= 0) return 0
  return Math.floor((200 * part + total) / (2 * total))
}

export interface BoardStats {
  total: number
  by_column: Record<string, number>
  done_column: string
  done: number
  progress: number
}

export function boardStats(board: Board): BoardStats {
  const byColumn: Record<string, number> = {}
  for (const column of board.columns) byColumn[column.id] = board.cards.filter((card) => card.column_id === column.id).length
  const known = board.cards.filter((card) => findColumn(board, card.column_id)).length
  const doneId = doneColumn(board).id
  const done = board.cards.filter((card) => card.column_id === doneId).length
  return { total: known, by_column: byColumn, done_column: doneId, done, progress: percent(done, known) }
}

export function checklistProgress(done: number, total: number): { done: number; total: number; percent: number } {
  return { done, total, percent: percent(done, total) }
}
